File: pattern_match_MP.cpp
/* 使用MP算法进行模式匹配 */

#include <cstring>
#include <iostream>

/*
 * 获取失效列表的过程也是一个模式匹配过程，需要是两个模式字符串进行相对平移来进行匹配。
 */
void calF(const char * s, int * F)
{
    int i = 0;
    int j = -1;
    F[0] = -1;

    while (s[i] != '\0')
    {
        while (j > -1 && s[i] != s[j])
        {
            // 利用前面的失效函数进行对齐
            j = F[j];
        }
        F[++i] = ++j;
    }
}

int MP_match(const char * target, const char * pattern)
{
    int i = 0, j = 0;
    int len_pattern = strlen(pattern);

    int * F = new int[len_pattern];

    // 获取失效列表
    calF(pattern, F);

    while (target[i] != '\0')
    {
        while (j >= 0 && target[i] != pattern[j])
        {
            // 平移模式串对齐
            j = F[j];
        }

        if (j >= len_pattern - 1)
        {
            return i - len_pattern + 1;
        }

        j++;
        i++;
    }

    delete [] F;

    return -1;
}

int main()
{
    char target[50];
    char pattern[50];

    std::cout << "Input target string:";
    std::cin >> target;

    std::cout << "Input patter string:";
    std::cin >> pattern;

    int result = MP_match(target, pattern);

    if (result < 0)
    {
        std::cout << "Not found." << std::endl;
    }
    else
    {
        std::cout << "Found pattern, index = " << result << std::endl;
    }

    return 0;
}

File: pattern_match_MP.py
# -*- coding: utf-8 -*-
# MP算法进行模式匹配的python实现

# 获取失效列表
def calF(s):
    F = [0]*(len(s) + 1)
    F[0] = -1
    i = 0
    j = -1

    while i < len(s):
        while j > -1 and s[i] != s[j]:
            j = F[j]
        j += 1
        i += 1
        F[i] = j

    return F

def match_MP(target, pattern):
    i = 0
    j = 0

    F = calF(pattern)

    while i < len(target):
        while j >= 0 and target[i] != pattern[j]:
            j = F[j]
        if j >= len(pattern) - 1:
            return i - len(pattern) + 1
        i += 1
        j += 1

    return -1

